pkg/apps/application-list.jsx

```diff
@@ -119,13 +124,35 @@ export const ApplicationList = ({ metainfo_db, appProgress, appProgressTitle, ac
         }
     }
 
+    const check_missing_data = useCallback(() => {
+        if (dataPackage !== null) {
+            setDataPackagesInstalled(true);
+            check_missing_packages(dataPackage)
+                    .then(data => setDataPackagesInstalled(data.missing_names?.length === 0 && data.unavailable_names?.length === 0));
+        }
+    }, [dataPackage]);
+
+    useInit(() => {
+        read_os_release().then(os_release => {
+            setDataPackage(get_config('appstream_data_packages', os_release, []));
+            setConfigPackage(get_config('appstream_config_packages', os_release, []));
+        }).catch(show_error);
+    });
+
+    useEffect(() => {
+        check_missing_data();
+    }, [dataPackage, check_missing_data]);
```

Comment:
TBH the use of all of the three use{Callback,Init,Effect} overwhelms my mind. Any one of them is prone to cause event loops, and all three of them is definitively too complicated. It's also strange to keep `appstream_{config,data}_packages` as state, as it's really not -- it won't change throughout the lifetime of the component, and even if it does, it does not actually update the UI (as your test case shows, it requires a reload).

I suggest to just use something like
```js
useInit(async () =>  {
    this.os_release_promise = read_os_release();
    const os_release = await this.os_release_promise;
    ... get_config() ...
    ... await check_missing_packages ...
   setDataPackagesInstalled()
}
```
this does it all in one go, and then refresh() can re-use `this.os_release_promise().then(...)`.

Reply:
Having trouble reusing `os_release_promise` within refresh without a proper state, any tips here? or we can go back to rereading read_os_release multiple times, I still need to rerun the check to make sure the banner can go away anyways.

Reply:
Yes, you can `.then()` a promise many times, from the second time on it'll be "free". That's why I suggested the above schema, then you don't need to re-read os-release in refresh().

Of course feel free to use promise-style in useInit() as well, just async is a bit nicer to read.

Reply:
Right that is not the bit I got confused of, I wasnt able to read os_release_promise from refresh, it got null

Reply:
Done that, I got hopefully everything done sans the reuse `os_release_promise` bit.